src/compiler/scala/tools/nsc/transform/async/AsyncPhase.scala

```diff
@@ -0,0 +1,290 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.tools.nsc.transform.async
+
+import scala.collection.mutable
+import scala.tools.nsc.transform.async.user.FutureSystem
+import scala.tools.nsc.transform.{Transform, TypingTransformers}
+import scala.reflect.internal.Flags
+import scala.reflect.io.AbstractFile
+
+abstract class AsyncPhase extends Transform with TypingTransformers with AnfTransform with AsyncAnalysis with Lifter with LiveVariables {
+  self =>
+  import global._
+
+  private[async] var currentTransformState: AsyncTransformState[global.type] = _
+  private[async] val asyncNames = new AsyncNames[global.type](global)
+  protected[async] val tracing = new Tracing
+
+  val phaseName: String = "async"
+  override def enabled: Boolean = settings.async
+
+  private final class FutureSystemAttachment(val system: FutureSystem) extends PlainAttachment
+
+  // Optimization: avoid the transform altogether if there are no async blocks in a unit.
+  private val units = perRunCaches.newSet[CompilationUnit]()
+  final def addFutureSystemAttachment(unit: CompilationUnit, method: Tree, system: FutureSystem): method.type = {
+    units += unit
```

Comment:
With the FutureSystem being attached to an `async`, it looks like there's noting to prevent an invalid `await` in a compilation unit that doesn't contain any `async` because the entire phase will be skipped.

Reply:
This just skips the tree transform to these units. We still run the traversal looking for naked await calls in all units if one or more of the registered `await` symbols has been initialized.

Reply:
Test in https://github.com/retronym/scala/pull/88/files#diff-0f2308f0c5d3d263c38e8f4b69b6d5b4